library/std/src/io/mod.rs

```diff
@@ -1382,7 +1385,13 @@ impl<'a> IoSlice<'a> {
         // Total length of all the to be removed buffers.
         let mut accumulated_len = 0;
         for buf in bufs.iter() {
-            if accumulated_len + buf.len() > n {
+            // This check is equivalent to `accumulated_len + buf.len() > n`.
+            // `bufs` may contain aliased slices. If those slices are large enough,
+            // they may cause `accumulated_len + buf.len()` to overflow. Moving
+            // `accumulated_len` to the right term of the equation avoids this
+            // situation. `accumulated_len` will never be greater than `n`, so the
+            // subtraction will not underflow.
```

Comment:
It seems like the thing we're actually interested in *is* `n - accumulated_len`. Maybe we should be computing it in this loop directly, with a condition of `if let Some(remainder) = left.checked_sub(buf.len())` or so, where `left` starts at `n`?

Reply:
Good idea!